ButtonComponents: Fixed parameter naming inconsistency between compon…

Rickcern44 opened the pull request:
…ents (#4043)

<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

The changes are to address bug #4043. I renamed the Link Attribute/Parameter to Href in all buttons. Updated the docs to match this change.

Just as an FYI the only thing that I was not able to change or misunderstood was the XML code snippets. They are still labeled with Link instead of Href. I was able to change them locally, but that folder is ignored so the changes will not show here. I searched the entire solution and was unable to find any other reference to them. If there is let me know and I can go and change them.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Visual | All unit tests passing

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->
![image](https://user-images.githubusercontent.com/17716893/156765894-7d416d56-8b01-4b96-ab40-a21c35ce4de1.png)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

JonBunator commented on src/MudBlazor/Base/MudBaseButton.cs:
```diff
@@ -35,7 +35,7 @@ public abstract class MudBaseButton : MudComponentBase
         [Parameter]
         [Category(CategoryTypes.Button.ClickAction)]
-        public string Link { get; set; }
+        public string Href { get; set; }
```
Can you please keep the old `Link` prop, add an `Obsolete` attribute to it and forward the value of the old prop to the new one. Thanks

JonBunator reviewed (changes requested):
Thanks for the PR!

JonBunator reviewed (changes requested):
The following components also have a Link prop instead of Href:
`MudChip`
`MudMenuItem`
Can you please also change the prop?

Rickcern44 commented:
@JonBunator I also added the Href update to the two Components that I missed.

JonBunator commented on src/MudBlazor/Base/MudBaseButton.cs:
```diff
@@ -35,7 +36,16 @@ public abstract class MudBaseButton : MudComponentBase
+        /// <summary>
+        /// If set to a URL, clicking the button will open the referenced document. Use Target to specify where (Obsolete replaced by Href)
+        /// </summary>
+        [Obsolete("Use Href Instead.", true)]
```
```suggestion
        [Obsolete("Use Href Instead.", false)]
```
We want to keep this set to `false` until the next major release for backwards compability.

JonBunator commented on src/MudBlazor/Components/Chip/MudChip.razor.cs:
```diff
@@ -160,7 +160,7 @@ private Color GetColor()
         [Parameter]
         [Category(CategoryTypes.Chip.ClickAction)]
-        public string Link { get; set; }
+        public string Href { get; set; }
```
Can you also add an obsolete attribute to Link, please? Like you did in `MudBaseButton`

JonBunator commented on src/MudBlazor/Components/Menu/MudMenuItem.razor.cs:
```diff
@@ -15,7 +15,7 @@ public partial class MudMenuItem : MudComponentBase
         [Inject] public IJsApiService JsApiService { get; set; }
 
-        [Parameter] [Category(CategoryTypes.Menu.ClickAction)] public string Link { get; set; }
+        [Parameter] [Category(CategoryTypes.Menu.ClickAction)] public string Href { get; set; }
```
Can you also add an obsolete attribute to Link, please? Like you did in `MudBaseButton`

JonBunator reviewed (changes requested):
Thanks for the update! The build pipeline seems to fail. Can you please take a look?

Rickcern44 commented:
Okay, apologies for the missed commit not sure how it happened but I made the changes and made sure everything built. I also changed the Obsolete to false and added the Obsolete tag to the other Components. Sorry for all the confusion I am pretty new to contributing. I will try and pay better attention in the future.

JonBunator reviewed (approved):
No worries. Thank you!

henon commented:
Guys, please follow our protocol in dealing with breaking API renames. Normally we keep the old parameter but mark it with the `[Obsolete]` attribute and forward it to the new renamed parameter. Then there is no breaking change really. 

Since this PR is scheduled to be released with 6.0.8 and it is breaking we need to make it non-breaking like described above. Otherwise we'll have to revert and release it together with other breaking changes in a 6.1.0

JonBunator commented:
> Guys, please follow our protocol in dealing with breaking API renames. Normally we keep the old parameter but mark it with the `[Obsolete]` attribute and forward it to the new renamed parameter. Then there is no breaking change really. 
> 
> Since this PR is scheduled to be released with 6.0.8 and it is breaking we need to make it non-breaking like described above. Otherwise we'll have to revert and release it together with other breaking changes in a 6.1.0

This change is non breking. He added the obsolete prop the old parameter. But he probably forgot to fix the test and the pipeline didn't fail

henon commented:
Yes sorry, I just saw the obsoleted `Link` property in Chip. But then why did the test that broke not work any more? I don't understand

JonBunator commented:
> Yes sorry, I just saw the obsoleted `Link` property in Chip. But then why did the test that broke not work any more? I don't understand

Because of `Treat warnings as errors`. And the pipeline didn't fail because the test was excluded (Ignore attribute)

henon commented:
No, because the Link parameter was added with out the `[Parameter]` attribute. I fill fix that

henon commented:
Now it is really non-breaking. Thanks @JonBunator and @Rickcern44

Rickcern44 commented:
@henon @JonBunator Sorry for the miss on that parameter. First time working with this stuff. In the future if I work on this project, I will be sure to double check things like that.
